diesel/src/query_source/joins.rs

```diff
@@ -106,41 +106,59 @@ impl<Left, Right> QuerySource for Join<Left, Right, Inner>
 where
     Left: QuerySource + AppendSelection<Right::DefaultSelection>,
     Right: QuerySource,
-    Left::Output: SelectableExpression<Self>,
+    Left::Output: AppearsOnTable<Self>,
     Self: Clone,
 {
     type FromClause = Self;
-    type DefaultSelection = Left::Output;
+    // combining two valid selectable expressions for both tables will always yield a
+    // valid selectable expressions for the whole join, so no need to check that here
+    // again. These checked turned out to be quite expensive in terms of compile time
+    // so we use a wrapper type to just skip the check and forward other more relevant
+    // trait implementations to the inner type
+    //
+    // See https://github.com/diesel-rs/diesel/issues/3223 for details
+    type DefaultSelection = self::private::SkipSelectableExpressionWrapper<Left::Output>;
```

Comment:
Interesting... How did you trick the compiler into disabling check? For me, this looks like an insoluble task

Reply:
The relevant line is here: https://github.com/diesel-rs/diesel/pull/3288/files#diff-d0e8a449e498610baa9a5f3086200726d02bd5e1612f5f1d5587c4418baddf0eR482
This impl does not check that the trait holds true for the inner type, so it effectively skips the recursive trait bound check on large tuples, which is the underlying problem.

Reply:
You mean, that the most important thing there that `T` is not bounded, so the compiler does not do expensive recursive checks as it did before?

Reply:
Yes that's the main optimization.